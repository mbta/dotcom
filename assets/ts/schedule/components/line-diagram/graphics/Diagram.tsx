import React, { ReactElement } from "react";
import { routeToModeName } from "../../../../helpers/css";
import {
  isBranchingNode,
  largestSliceSize,
  longestPath,
  longestPathLength,
  nextStopIds,
  stopForId,
  stopIds
} from "../../../../helpers/stop-tree";
import { isAGreenLineRoute } from "../../../../models/route";
import { Alert, DirectionId, Route } from "../../../../__v3api";
import {
  IndexedRouteStop,
  RouteStop,
  StopId,
  StopTree
} from "../../__schedule";
import { diagramWidth } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";
import Stop from "./Stop";
import { Line, SimpleLine } from "./Line";
import Merges from "./Merges";

interface Props {
  stopTree: StopTree;
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const routeName = (route: Route): string =>
  isAGreenLineRoute(route) ? "Green Line" : route.name;

const diagramDescription = (
  pathLength: number,
  route: Route,
  directionId: DirectionId
): string => {
  const text = `${pathLength} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = route;
  if (destinations) {
    return `${text} going ${names[directionId]} to ${destinations[directionId]}`;
  }
  return text;
};

const hasBranchLines = (stopTree: StopTree): boolean =>
  largestSliceSize(stopTree) > 1;

const branchingDescription = (stopTree: StopTree): string => {
  const primaryPath = longestPath(stopTree);
  const branchingStops: StopId[] = primaryPath.filter(id =>
    isBranchingNode(stopTree, id)
  );
  const branchText = branchingStops.map(branchingStopId => {
    const routeStop: RouteStop = stopForId(stopTree, branchingStopId);
    const branchNames: string[] = nextStopIds(stopTree, branchingStopId)
      .filter(id => !primaryPath.includes(id))
      .map(id => stopForId(stopTree, id)?.branch || "");

    return `Use ${routeStop.name} to transfer to trains on ${branchNames.join(
      " or "
    )} branches.`;
  });
  return branchText.join("");
};

interface SimpleProps {
  routeStopList: IndexedRouteStop[];
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const SimpleDiagram = ({
  routeStopList,
  route,
  directionId,
  alerts
}: SimpleProps): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(routeStopList.length, route, directionId)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>
      {/* Draw lines between stops */
      routeStopList.map((routeStop, index) => {
        const nextStop = routeStopList[index + 1];
        if (!nextStop) return null;
        return (
          <SimpleLine
            key={`${routeStop.id}-${nextStop.id}`}
            fromId={routeStop.id}
            toId={nextStop.id}
            alerts={alerts}
          />
        );
      })}

      {/* Draw circles for each stop */
      routeStopList.map(routeStop => (
        <Stop key={`stop-${routeStop.id}`} stopId={routeStop.id} />
      ))}
    </svg>
  </>
);

const Diagram = ({
  stopTree,
  route,
  directionId,
  alerts
}: Props): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(longestPathLength(stopTree), route, directionId)}
        {hasBranchLines(stopTree) && branchingDescription(stopTree)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>

      {/* If there are multiple branches, draw the lines and curves for those */
      hasBranchLines(stopTree) && (
        <Merges stopTree={stopTree} alerts={alerts} />
      )}

      {/* Draw lines between stops */
      stopIds(stopTree).map(stopId =>
        nextStopIds(stopTree, stopId).map(nextStopId => (
          <Line
            key={`${stopId}-${nextStopId}`}
            stopTree={stopTree}
            fromId={stopId}
            toId={nextStopId}
            alerts={alerts}
          />
        ))
      )}

      {/* Draw circles for each stop */
      stopIds(stopTree).map(stopId => (
        <Stop key={stopId} stopId={stopId} />
      ))}
    </svg>
  </>
);

export { Diagram, SimpleDiagram };
