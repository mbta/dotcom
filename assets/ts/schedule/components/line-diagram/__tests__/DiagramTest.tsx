import React from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import { RouteStop, StopTree } from "../../__schedule";
import { createStopTreeCoordStore } from "../graphics/useTreeStopPositions";
import { Diagram } from "../graphics/Diagram";
import { Route, RouteType } from "../../../../__v3api";
import Stop from "../graphics/Stop";
import { LiveDataByStop } from "../__line-diagram";

const stopTree: StopTree = {
  byId: {
    a1: { id: "a1", value: { id: "a1" } as RouteStop },
    a2: { id: "a2", value: { id: "a2" } as RouteStop },
    b1: { id: "b1", value: { id: "b1" } as RouteStop },
    b2: { id: "b2", value: { id: "b2" } as RouteStop },
    b3: { id: "b3", value: { id: "b3" } as RouteStop },
    c1: { id: "c1", value: { id: "c1" } as RouteStop },
    c2: { id: "c2", value: { id: "c2" } as RouteStop },
    m1: { id: "m1", value: { id: "m1" } as RouteStop },
    m2: { id: "m2", value: { id: "m2" } as RouteStop },
    m3: { id: "m3", value: { id: "m3" } as RouteStop },
    x1: { id: "x1", value: { id: "x1" } as RouteStop },
    x2: { id: "x2", value: { id: "x2" } as RouteStop },
    y1: { id: "y1", value: { id: "y1" } as RouteStop }
  },
  edges: {
    a1: { next: ["a2"], previous: [] },
    a2: { next: ["m1"], previous: ["a1"] },
    b1: { next: ["b2"], previous: [] },
    b2: { next: ["b3"], previous: ["b1"] },
    b3: { next: ["m1"], previous: ["b2"] },
    c1: { next: ["c2"], previous: [] },
    c2: { next: ["m2"], previous: ["c1"] },
    m1: { next: ["m2"], previous: ["a2", "b3"] },
    m2: { next: ["m3"], previous: ["c2", "m1"] },
    m3: { next: ["x1", "y1"], previous: ["m2"] },
    x1: { next: ["x2"], previous: ["m3"] },
    x2: { next: [], previous: ["x1"] },
    y1: { next: [], previous: ["m3"] }
  },
  startingNodes: ["a1", "b1", "c1"]
};
const store = createStopTreeCoordStore(stopTree);

const route: Route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "frequent_bus_route",
  line_id: null
};

describe("Diagram", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <Diagram
          stopTree={stopTree}
          route={route}
          directionId={1}
          alerts={[]}
        />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("uses the route color CSS class", () => {
    expect(wrapper.exists(".line-diagram-svg.bus")).toBeTruthy();
  });

  it("shows an SVG", () => {
    expect(wrapper.exists("svg.line-diagram-svg")).toBeTruthy();
    expect(wrapper.exists("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("should draw each stop", () => {
    expect(wrapper.exists(Stop)).toBeTruthy();
    expect(wrapper.find(Stop)).toHaveLength(13);
  });

  it("shows merges for branches", () => {
    expect(wrapper.exists("g.line-diagram-svg__merge")).toBeTruthy();
  });
});
