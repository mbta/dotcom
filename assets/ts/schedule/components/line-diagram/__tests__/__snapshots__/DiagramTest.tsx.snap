// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <Diagram stopTree={{...}} route={{...}} directionId={1} alerts={{...}} liveData={{...}}>
    <svg xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" aria-labelledby=\\"diagram-title diagram-desc\\" className=\\"line-diagram-svg bus\\" width=\\"36px\\" height=\\"100%\\" style={{...}}>
      <title id=\\"diagram-title\\">
        Line diagram for 
        route 1
      </title>
      <desc id=\\"diagram-desc\\">
        8 stops going Inbound to End
        Use undefined to transfer to trains on  branches.
      </desc>
      <defs>
        <pattern id=\\"diagonalHatch\\" width=\\"10\\" height=\\"10\\" patternTransform=\\"rotate(45 0 0)\\" patternUnits=\\"userSpaceOnUse\\">
          <line className=\\"line-diagram-svg__line\\" y2={8} strokeWidth=\\"16px\\" />
        </pattern>
      </defs>
      <Merges stopTree={{...}} alerts={{...}}>
        <g className=\\"line-diagram-svg__merge\\">
          <Line stopTree={{...}} fromId=\\"a2\\" toId=\\"m1\\" alerts={{...}} />
          <Line stopTree={{...}} fromId=\\"c2\\" toId=\\"m2\\" alerts={{...}} />
          <Line stopTree={{...}} fromId=\\"m3\\" toId=\\"y1\\" alerts={{...}} />
        </g>
      </Merges>
      <Line stopTree={{...}} fromId=\\"a1\\" toId=\\"a2\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"a2\\" toId=\\"m1\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"b1\\" toId=\\"b2\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"b2\\" toId=\\"b3\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"b3\\" toId=\\"m1\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"c1\\" toId=\\"c2\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"c2\\" toId=\\"m2\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"m1\\" toId=\\"m2\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"m2\\" toId=\\"m3\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"m3\\" toId=\\"x1\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"m3\\" toId=\\"y1\\" alerts={{...}} />
      <Line stopTree={{...}} fromId=\\"x1\\" toId=\\"x2\\" alerts={{...}} />
      <Stop stopId=\\"a1\\" />
      <Stop stopId=\\"a2\\" />
      <Stop stopId=\\"b1\\" />
      <Stop stopId=\\"b2\\" />
      <Stop stopId=\\"b3\\" />
      <Stop stopId=\\"c1\\" />
      <Stop stopId=\\"c2\\" />
      <Stop stopId=\\"m1\\" />
      <Stop stopId=\\"m2\\" />
      <Stop stopId=\\"m3\\" />
      <Stop stopId=\\"x1\\" />
      <Stop stopId=\\"x2\\" />
      <Stop stopId=\\"y1\\" />
    </svg>
  </Diagram>
</Provider>"
`;
